### PR TITLE
Add mesh shader support: Support fragment shader per-primitive input

### PR DESCRIPTION
**Pull request by amdrexu**
Add a 'perPrimitive' flag to fragment shader input import call. The
attribute locations for per-primitive inputs of fragment shader follow
those per-vertex inputs. This layout is required by HW.

Note that the emulated built-in input, such as ViewportIndex,
PrimitiveId, Layer, ViewIndex, are mapped to per-primitive built-ins for
fragment shader if the previous stage is mesh shader. This is different
from traditional graphics pipeline.

Also, add 'perPrimitive' field to interpolation info. The field will be
used to program the register field SPI_PS_IN_CONTRO.NUM_PRIM_INTERP.
A later PR will shows the usage.

### CODE REVIEW

File: lgc/patch/PatchInOutImportExport.cpp [hunk 1]
```diff
@@ -2553,16 +2575,22 @@ Value *PatchInOutImportExport::patchFsBuiltInInputImport(Type *inputTy, unsigned
...
 
     // Emulation for "in int gl_PrimitiveID" or "in int gl_Layer" or "in int gl_ViewportIndex"
     // or "in int gl_ViewIndex"
-    input = patchFsGenericInputImport(inputTy, loc, nullptr, nullptr, nullptr, InOutInfo::InterpModeFlat,
-                                      InOutInfo::InterpLocCenter, false, insertPos);
+    bool isPerPrimitive = false;
+    if (m_pipelineState->getPrevShaderStage(ShaderStageFragment) == ShaderStageMesh) {
+      // NOTE: If the previous shader stage is mesh shader, those built-ins are exported via primitive attributes.
+      isPerPrimitive = true;
+    }
```
Comment (nhaehnle) [inline lgc/patch/PatchInOutImportExport.cpp:2591]: You could set `isPerPrimitive` in the if/else above.

**Review by nhaehnle — COMMENTED** [review]
Thanks, looks good to me but I do have a minor suggestion inline.

**Comment by qiaojbao** [comment]
[42368/73410] Pass: 17999 Fail: 7 Skip: 15148 ExpectedFail: 0 UnexpectedPass: 0 Crash: 3688 Timeout: 0 Missing: 5526 Flake: 0 Duration: 20:56 Remaining: 15:20
Test timeout.

File: lgc/patch/PatchInOutImportExport.cpp (hunk 1 above)
Comment (amdrexu) [reply lgc/patch/PatchInOutImportExport.cpp:2591]: OK. Will be updated.

**Comment by amdrexu** [comment]
> [42368/73410] Pass: 17999 Fail: 7 Skip: 15148 ExpectedFail: 0 UnexpectedPass: 0 Crash: 3688 Timeout: 0 Missing: 5526 Flake: 0 Duration: 20:56 Remaining: 15:20 Test timeout.

The failure is caused by a latent issue. Fix by https://github.com/GPUOpen-Drivers/llpc/pull/1945

**Comment by amdvlk-admin** [comment]
### Test summary for commit d67a4acce2e3d64b14be259b9ef3b5e69b55971a
<details><summary>CTS tests (Failed: 0/216059)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35298/73410 (48.1%)</li>
<li>Failed: 0/73410 (0.0%)</li>
<li>Not Supported: 38112/73410 (51.9%)</li><li>Warnings: 0/73410 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33017/69239 (47.7%)</li>
<li>Failed: 0/69239 (0.0%)</li>
<li>Not Supported: 36222/69239 (52.3%)</li><li>Warnings: 0/69239 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73410 (51.2%)</li>
<li>Failed: 0/73410 (0.0%)</li>
<li>Not Supported: 35857/73410 (48.8%)</li><li>Warnings: 0/73410 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 93dfd58bfa583b2a27762b70d06f5ec77874afc4
<details><summary>CTS tests (Failed: 0/215432)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35298/73201 (48.2%)</li>
<li>Failed: 0/73201 (0.0%)</li>
<li>Not Supported: 37903/73201 (51.8%)</li><li>Warnings: 0/73201 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33017/69030 (47.8%)</li>
<li>Failed: 0/69030 (0.0%)</li>
<li>Not Supported: 36013/69030 (52.2%)</li><li>Warnings: 0/69030 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73201 (51.3%)</li>
<li>Failed: 0/73201 (0.0%)</li>
<li>Not Supported: 35648/73201 (48.7%)</li><li>Warnings: 0/73201 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit efb5c7d9c89e6d6c43bf6f2cca1e404468c77b3a
<details><summary>CTS tests (Failed: 1/215432)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35298/73201 (48.2%)</li>
<li>Failed: 0/73201 (0.0%)</li>
<li>Not Supported: 37903/73201 (51.8%)</li><li>Warnings: 0/73201 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33016/69030 (47.8%)</li>
<li><details><summary>Failed: 1/69030 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.binding.priority.regular.dedicated.buffer_1048577
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 36013/69030 (52.2%)</li><li>Warnings: 0/69030 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73201 (51.3%)</li>
<li>Failed: 0/73201 (0.0%)</li>
<li>Not Supported: 35648/73201 (48.7%)</li><li>Warnings: 0/73201 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7ae6d036310323b8a965ec623c0499648afb1f8e
<details><summary>CTS tests (Failed: 3/215435)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35298/73202 (48.2%)</li>
<li><details><summary>Failed: 1/73202 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.graphicsfuzz.cov-instruction-simplify-atanh-log-undefined
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 37903/73202 (51.8%)</li><li>Warnings: 0/73202 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33017/69031 (47.8%)</li>
<li><details><summary>Failed: 1/69031 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.graphicsfuzz.cov-instruction-simplify-atanh-log-undefined
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 36013/69031 (52.2%)</li><li>Warnings: 0/69031 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73202 (51.3%)</li>
<li><details><summary>Failed: 1/73202 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.graphicsfuzz.cov-instruction-simplify-atanh-log-undefined
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 35648/73202 (48.7%)</li><li>Warnings: 0/73202 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit b69d7442ae44a0faf8a861daa849efca939f31a0
<details><summary>CTS tests (Failed: 1/215435)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 0/73202 (0.0%)</li>
<li>Failed: 0/73202 (0.0%)</li>
<li>Not Supported: 73202/73202 (100.0%)</li><li>Warnings: 0/73202 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 0/69031 (0.0%)</li>
<li>Failed: 0/69031 (0.0%)</li>
<li>Not Supported: 69031/69031 (100.0%)</li><li>Warnings: 0/69031 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73202 (51.3%)</li>
<li><details><summary>Failed: 1/73202 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.graphicsfuzz.cov-instruction-simplify-atanh-log-undefined
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 35648/73202 (48.7%)</li><li>Warnings: 0/73202 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit b69d7442ae44a0faf8a861daa849efca939f31a0
<details><summary>CTS tests (Failed: 1/215435)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 0/73202 (0.0%)</li>
<li>Failed: 0/73202 (0.0%)</li>
<li>Not Supported: 73202/73202 (100.0%)</li><li>Warnings: 0/73202 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 0/69031 (0.0%)</li>
<li>Failed: 0/69031 (0.0%)</li>
<li>Not Supported: 69031/69031 (100.0%)</li><li>Warnings: 0/69031 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73202 (51.3%)</li>
<li><details><summary>Failed: 1/73202 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.graphicsfuzz.cov-instruction-simplify-atanh-log-undefined
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 35648/73202 (48.7%)</li><li>Warnings: 0/73202 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by qiaojbao** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit b69d7442ae44a0faf8a861daa849efca939f31a0
<details><summary>CTS tests (Failed: 1/215435)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 0/73202 (0.0%)</li>
<li>Failed: 0/73202 (0.0%)</li>
<li>Not Supported: 73202/73202 (100.0%)</li><li>Warnings: 0/73202 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 0/69031 (0.0%)</li>
<li>Failed: 0/69031 (0.0%)</li>
<li>Not Supported: 69031/69031 (100.0%)</li><li>Warnings: 0/69031 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73202 (51.3%)</li>
<li><details><summary>Failed: 1/73202 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.graphicsfuzz.cov-instruction-simplify-atanh-log-undefined
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 35648/73202 (48.7%)</li><li>Warnings: 0/73202 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]
Thanks, LGTM

**Comment by qiaojbao** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit b69d7442ae44a0faf8a861daa849efca939f31a0
<details><summary>CTS tests (Failed: 0/214796)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 0/72989 (0.0%)</li>
<li>Failed: 0/72989 (0.0%)</li>
<li>Not Supported: 72989/72989 (100.0%)</li><li>Warnings: 0/72989 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 0/68818 (0.0%)</li>
<li>Failed: 0/68818 (0.0%)</li>
<li>Not Supported: 68818/68818 (100.0%)</li><li>Warnings: 0/68818 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/72989 (51.5%)</li>
<li>Failed: 0/72989 (0.0%)</li>
<li>Not Supported: 35436/72989 (48.5%)</li><li>Warnings: 0/72989 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**